File: src/handlers/users/provider/facebook.rs
use crate::{
    config::Config,
    handlers::users::provider::{Provider, UserProvidedData},
};
use reqwest::{blocking::Client, Error};
use serde::Deserialize;

#[derive(Deserialize)]
struct PictureData {
    pub is_silhouette: bool,
    pub url: String,
}

#[derive(Deserialize)]
struct Data {
    pub data: PictureData,
}

#[derive(Deserialize)]
struct FacebookUser {
    pub email: Option<String>,
    pub name: Option<String>,
    pub picture: Data,
}

#[derive(Clone)]
pub struct FacebookProvider {
    config: Config,
}

impl FacebookProvider {
    pub fn new(config: Config) -> FacebookProvider {
        return FacebookProvider { config };
    }
}

impl Provider for FacebookProvider {
    fn enabled(&self) -> bool {
        return self.config.facebook_enabled;
    }

    fn auth_url(&self) -> String {
        return String::from("https://www.facebook.com/v5.0/dialog/oauth");
    }

    fn token_url(&self) -> String {
        return String::from("https://graph.facebook.com/v5.0/oauth/access_token");
    }

    fn client_id(&self) -> String {
        return self.config.facebook_client_id.clone().unwrap();
    }

    fn client_secret(&self) -> String {
        return self.config.facebook_client_secret.clone().unwrap();
    }

    fn scopes(&self) -> Vec<String> {
        return vec![String::from("email")];
    }

    fn get_user_data(&self, access_token: String) -> Result<UserProvidedData, Error> {
        let client = Client::new();

        let response = client
            .get("https://graph.facebook.com/me?fields=name,email,picture{url,is_silhouette}")
            .bearer_auth(access_token)
            .send()?;

        let data: FacebookUser = response.json()?;

        let d = UserProvidedData {
            email: data.email,
            verified: true,
            name: data.name,
            avatar: if !data.picture.data.is_silhouette { Some(data.picture.data.url) } else { None },
        };

        return Ok(d);
    }
}

File: handlers/provider/facebook.go
package provider

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/zolamk/trust/config"
	"golang.org/x/oauth2"
	"golang.org/x/oauth2/facebook"
)

type PictureData struct {
	IsSilhouette bool   `json:"is_silhouette"`
	URL          string `json:"url"`
}

type Data struct {
	Data PictureData `json:"data"`
}

type FacebookUser struct {
	Email   *string `json:"email"`
	Name    *string `json:"name"`
	Picture Data    `json:"picture"`
}

type FacebookProvider struct {
	config *config.Config
}

func (f *FacebookProvider) name() string {
	return "facebook"
}

func (f *FacebookProvider) enabled() bool {
	return f.config.Facebook.Enabled
}

func (f *FacebookProvider) get_config() *oauth2.Config {
	return &oauth2.Config{
		ClientID:     f.config.Facebook.ID,
		ClientSecret: f.config.Facebook.Secret,
		Scopes:       []string{"email"},
		Endpoint:     facebook.Endpoint,
		RedirectURL:  fmt.Sprintf("%s/authorize/callback", f.config.InstanceURL),
	}
}

func (f *FacebookProvider) get_user_data(access_token string) (*UserData, error) {

	req, err := http.NewRequest("GET", "https://graph.facebook.com/me?fields=name,email,picture{url,is_silhouette}", nil)

	if err != nil {
		return nil, err
	}

	req.Header.Add("authorization", fmt.Sprintf("Bearer %s", access_token))

	res, err := http.DefaultClient.Do(req)

	if err != nil {
		return nil, err
	}

	defer res.Body.Close()

	decoder := json.NewDecoder(res.Body)

	data := &FacebookUser{}

	if err := decoder.Decode(data); err != nil {
		return nil, err
	}

	user_data := &UserData{
		Name:     data.Name,
		Email:    data.Email,
		Verified: true,
	}

	if !data.Picture.Data.IsSilhouette {
		user_data.Avatar = &data.Picture.Data.URL
	}

	return user_data, nil

}
